service/worker/batcher/workflow.go

```diff
@@ -20,7 +20,328 @@
+}
+
+func setDefaultParams(params BatchParams) (BatchParams, error) {
+	if params.BatchType == "" || params.Reason == "" || params.DomainName == "" || params.QueryCondition == "" {
+		return BatchParams{}, fmt.Errorf("must provide required parameters: BatchType/Reason/DomainName/QueryCondition")
+	}
+	if params.RPS <= 0 {
+		params.RPS = defaultRPS
+	}
+	if params.Concurrency <= 0 {
+		params.Concurrency = defaultConcurrency
+	}
+	if params.AttemptsOnRetryableError <= 0 {
+		params.AttemptsOnRetryableError = defaultAttemptsOnRetryableError
+	}
+	if params.ActivityHeartBeatTimeout <= 0 {
+		params.ActivityHeartBeatTimeout = defaultActivityHeartBeatTimeout
+	}
+	switch params.BatchType {
+	case BatchTypeReset:
+		fallthrough
+	default:
+		return BatchParams{}, fmt.Errorf("not supported batch type: %v", params.BatchType)
+	case BatchTypeTerminate:
+
+	}
+	return params, nil
+}
+
+func BatchActivity(ctx context.Context, batchParams BatchParams) error {
+	batcher := ctx.Value(batcherContextKey).(*Batcher)
+
+	hbd := HeartBeatDetails{}
+	startOver := true
+	if activity.HasHeartbeatDetails(ctx) {
+		if err := activity.GetHeartbeatDetails(ctx, &hbd); err == nil {
+			startOver = false
+		} else {
+			//TODO error metrics
+			getActivityLogger(ctx).Error("Failed to recover from last heartbeat, start over from beginning", tag.Error(err))
+		}
+	}
+
+	if startOver {
+		resp, err := batcher.svcClient.CountWorkflowExecutions(ctx, &shared.CountWorkflowExecutionsRequest{
+			Domain: common.StringPtr(batchParams.DomainName),
+			Query:  common.StringPtr(batchParams.QueryCondition),
+		})
+		if err != nil {
+			return err
+		}
+		hbd.TotalEstimate = resp.GetCount()
+	}
+	rateLimiter := rate.NewLimiter(rate.Limit(batchParams.RPS), batchParams.RPS)
+	taskCh := make(chan taskDetail, pageSize)
+	var succCount, errCount int32
+	for i := 0; i < batchParams.Concurrency; i++ {
+		go startTaskProcessor(ctx, batchParams, taskCh, rateLimiter, &succCount, &errCount)
+	}
+
+	for {
+		// TODO https://github.com/uber/cadence/issues/2154
+		//  Need to improve scan concurrency because it will hold a ES resource for a long time.
+		//  we can't use list API because terminate / reset will mutate the result.
+		resp, err := batcher.svcClient.ScanWorkflowExecutions(ctx, &shared.ListWorkflowExecutionsRequest{
+			PageSize:      common.Int32Ptr(int32(pageSize)),
+			Domain:        common.StringPtr(batchParams.DomainName),
+			NextPageToken: hbd.PageToken,
+			Query:         common.StringPtr(batchParams.QueryCondition),
+		})
+		if err != nil {
+			return err
+		}
+
+		// clear all counters
+		atomic.StoreInt32(&succCount, 0)
+		atomic.StoreInt32(&errCount, 0)
+
+		// send all tasks
+		for _, wf := range resp.Executions {
+			taskCh <- taskDetail{
+				execution: *wf.Execution,
+				attempts:  0,
+				hbd:       hbd,
+			}
+		}
+		batchCount := len(resp.Executions)
+
+		// wait for counters indicate this batch is done
+	Loop:
+		for {
+			select {
```

Comment:
why counters? Why not just loop here waiting for responses with error or not.. That seems more useful - in case you want to return back errors/persist them somewhere.. 

I dont think this helps with performance also as the different go routines will have to contend on the atomic increment.

Reply:
The most cost is I/O to cadence server and ratelimiter and most of them times they are waiting for them. Because we have to use multiple go routines, waiting for return will not work. Also I wan to return counters to do heartbeat for visibility of the progress. Without these atomic counters, we can use channel to return counters. Or do you have any preference?

Reply:
As discussed, let's wait on a response channel for tasks to complete and accumulate error/success counts in this method.

Reply:
Done